### PR TITLE
[MM] Make controlnet image processors use MM cache system

### PR DESCRIPTION
**Pull request by lstein**
## Summary

This PR changes the loading of most controlnet image processors to use the model manager caching system. This should result in a performance improvement when multiple different control images are processed using the same processor. However, the application of the processed image during image generation will not be affected.

A couple of notes here.

1. The `controlnet_aux` module, which defines the image processors, uses an odd `from_pretrained()` pattern in which the same HuggingFace repo is used for each of the processor classes. Internally, each processor class looks for a different hard-coded model checkpoint file. To use the MM caching system, I figured out which model was being loaded, and created a mapping from the processor class to the checkpoint file in `controlnet_image_processors.py`. If at some point one of the processors in `controlnet_aux` is updated to use a different file name, a compensatory change will need to be made to `controlnet_image_processors.py`.
2. I had to make a small change to the model installer's `download_and_cache_model()` method in order to accommodate the way that the SegmentAnything class looks for its model in a subfolder.
3. I was unable to migrate the Leres processor due to the fact that its `from_pretrained()` method loads two different models, but `download_and_cache_model()` doesn't accept more than a single model file at a time.

## QA Instructions

In the workflow editor, execute the following controlnet processors and confirm that they load and work as expected:
* Midas Depth Processor
* MLSD Processor
* Pidi Processor
* Zoe (Depth) Processor
* Leres (Depth) Processor
* Segment Anything Processor
* Lineart Processor
* Lineart Anime Processor
* Hed Processor

The remaining processors (canny, shuffle, etc) do not download models from HF.

## Merge Plan

Merge when approved.

## Checklist

- [X] _The PR has a short but descriptive title, suitable for a changelog_
- [X] _Tests added / updated (if applicable)_
- [X] _Documentation added / updated (if applicable)_

### CODE REVIEW

**Review by RyanJDick — COMMENTED** [review]
Overall, makes sense to me. A few comments:

- I smoke tested the lineart processor. I trust that you have tested them all?
- The lineart processor (and probably others) are treated as having size 0 by `calc_model_size_by_data(...)`. Once you merge main into your branch, this will result in error logs (as of https://github.com/invoke-ai/InvokeAI/pull/6571).
- The lineart processor (and probably others) don't run on the GPU, because they don't have a `.to(...)` method. Is that expected?
- @brandonrising can you review for potential commercial implications?

**Comment by psychedelicious** [comment]
This is a good step in the right direction. Eventually we should finish the move away from `controlnet_aux` as it is not actively maintained.

File: invokeai/app/services/shared/invocation_context.py [hunk 1]
```diff
@@ -447,6 +447,7 @@ def search_by_attrs(
     def download_and_cache_model(
         self,
         source: str | AnyHttpUrl,
+        preserve_subfolders: bool = False,
```
[the remark this answers is not in the document]
Comment (lstein) [reply invokeai/app/services/shared/invocation_context.py:450]: Added.

File: invokeai/app/invocations/controlnet_image_processors.py [hunk 2]
```diff
@@ -132,6 +133,14 @@ class ImageProcessorInvocation(BaseInvocation, WithMetadata, WithBoard):
 
     image: ImageField = InputField(description="The image to process")
 
+    # Map controlnet_aux detector classes to model files in "lllyasviel/Annotators"
+    CONTROLNET_PROCESSORS: ClassVar[Dict[type, Path]] = {
+        MidasDetector: Path("dpt_hybrid-midas-501f0c75.pt"),
+        MLSDdetector: Path("mlsd_large_512_fp32.pth"),
+        PidiNetDetector: Path("table5_pidinet.pth"),
+        ZoeDetector: Path("ZoeD_M12_N.pt"),
+    }
```
[the remark this answers is not in the document]
Comment (lstein) [reply invokeai/app/invocations/controlnet_image_processors.py:142]: Done.

**Comment by lstein** [comment]
> Overall, makes sense to me. A few comments:
> 
> * I smoke tested the lineart processor. I trust that you have tested them all?

Yes, all tested successfully.

> * The lineart processor (and probably others) are treated as having size 0 by `calc_model_size_by_data(...)`. Once you merge main into your branch, this will result in error logs (as of [Fix some models treated as having size 0 in the model cache #6571](https://github.com/invoke-ai/InvokeAI/pull/6571)).

I've added memory calculations for the controlnet processors as well as the scheduler and tokenizer. Hopefully we won't see any more complaints about size 0.

> * The lineart processor (and probably others) don't run on the GPU, because they don't have a `.to(...)` method. Is that expected?

Some of the controlnet processors run in GPU and others in CPU. The CPU models only runs once when the encoded controlnet image is first calculated and cached, so it's not a high priority to fix at the moment.

> * @brandonrising can you review for potential commercial implications?

**Comment by brandonrising** [comment]
Screenshot 2024-07-23 at 3 29 04 PM
Hmm I did get this error when testing each of the controlnets the first time, but I don't seem to be able to reproduce it again.

Otherwise, this all worked really well and commercial would be improved by having these in the cache

**Comment by brandonrising** [comment]
> Screenshot 2024-07-23 at 3 29 04 PM Hmm I did get this error when testing each of the controlnets the first time, but I don't seem to be able to reproduce it again.
> 
> Otherwise, this all worked really well and commercial would be improved by having these in the cache

oh ignore that error message, I thought it was downloading a preprocessor but it was probably just from when I kicked off installs in MM

**Review by brandonrising — APPROVED** [review]

File: invokeai/backend/model_manager/load/model_cache/model_cache_default.py [hunk 3]
```diff
@@ -292,7 +292,13 @@ def move_model_to_device(self, cache_entry: CacheRecord[AnyModel], target_device
...
                         new_dict[k] = v.to(target_device, copy=True)
                     cache_entry.model.load_state_dict(new_dict, assign=True)
-            cache_entry.model.to(target_device)
+            try:
+                cache_entry.model.to(target_device)
+            except TypeError as e:
+                if "got an unexpected keyword argument 'non_blocking'" in str(e):
+                    cache_entry.model.to(target_device)
+                else:
+                    raise e
```
Comment (RyanJDick) [inline invokeai/backend/model_manager/load/model_cache/model_cache_default.py:301]: nit: Now that we removed the use of non-blocking on main (https://github.com/invoke-ai/InvokeAI/pull/6624), we could remove this try-except.

**Review by RyanJDick — APPROVED** [review]

**Comment by psychedelicious** [comment]
Thanks @lstein, I picked up on this for the Canvas v2 release and got all the nodes using the MM cache in #6831. Well, technically they don't exactly use the cache bc I wasn't sure how to get them to report on their sizes, but we have that on the TODO list.
